packages/amplify-cli-core/src/feature-flags/featureFlagProvider.ts

```diff
@@ -0,0 +1,422 @@
+import Ajv, { AdditionalPropertiesParams } from 'ajv';
+import * as fs from 'fs-extra';
+import _ from 'lodash';
+import * as path from 'path';
+import { JSONSchema7, JSONSchema7Definition } from 'json-schema';
+import { CLIEnvironmentProvider, JSONUtilities, JSONValidationError } from '..';
+import { FeatureFlags, FeatureFlagRegistration, FeatureFlagsEntry, FeatureFlagType } from '.';
+import { AmplifyConfigFileName } from '../constants';
+import { FeatureFlagFileProvider } from './featureFlagFileProvider';
+import { FeatureFlagEnvironmentProvider } from './featureFlagEnvironmentProvider';
+
+export class FeatureFlagProvider {
+  private static instance: FeatureFlagProvider;
+
+  private readonly registrations: Map<string, FeatureFlagRegistration[]> = new Map();
+  private effectiveFlags: Readonly<FeatureFlagsEntry> = {};
+  private newProjectDefaults: Readonly<FeatureFlagsEntry> = {};
+  private existingProjectDefaults: Readonly<FeatureFlagsEntry> = {};
+
+  private constructor(private environmentProvider: CLIEnvironmentProvider, private projectPath: string) {}
+
+  public static initialize = async (environmentProvider: CLIEnvironmentProvider, projectPath: string): Promise<FeatureFlagProvider> => {
```

Comment:
Is this class intended to be a singleton? (it seems like it should be) If so, this method needs to check if instance is already set.